code_generation/producers/ditaupairquantities.py

```diff
@@ -0,0 +1,179 @@
+import code_generation.quantities as q
+from code_generation.producer import Producer, ProducerGroup
+
+####################
+# Set of general producers for DiTauPair Quantities
+####################
+
+pt_1 = Producer(
+    name="pt_1",
+    call="quantities::pt({df}, {output}, {input})",
+    input=[q.p4_1],
+    output=[q.pt_1],
+    scopes=["mt", "et", "tt", "em"],
+)
+pt_2 = Producer(
+    name="pt_2",
+    call="quantities::pt({df}, {output}, {input})",
+    input=[q.p4_2],
+    output=[q.pt_2],
+    scopes=["mt", "et", "tt", "em"],
+)
+eta_1 = Producer(
+    name="eta_1",
+    call="quantities::eta({df}, {output}, {input})",
+    input=[q.p4_1],
+    output=[q.eta_1],
+    scopes=["mt", "et", "tt", "em"],
+)
+eta_2 = Producer(
+    name="eta_2",
+    call="quantities::eta({df}, {output}, {input})",
+    input=[q.p4_2],
+    output=[q.eta_2],
+    scopes=["mt", "et", "tt", "em"],
+)
+phi_1 = Producer(
+    name="phi_1",
+    call="quantities::phi({df}, {output}, {input})",
+    input=[q.p4_1],
+    output=[q.phi_1],
+    scopes=["mt", "et", "tt", "em"],
+)
+phi_2 = Producer(
+    name="phi_2",
+    call="quantities::phi({df}, {output}, {input})",
+    input=[q.p4_2],
+    output=[q.phi_2],
+    scopes=["mt", "et", "tt", "em"],
+)
+m_vis = Producer(
+    name="m_vis",
+    call="quantities::m_vis({df}, {output}, {input_vec})",
+    input=[q.p4_1, q.p4_2],
+    output=[q.m_vis],
+    scopes=["mt", "et", "tt", "em"],
+)
+####################
+# Set of channel specific producers
+####################
+mass_1 = Producer(
+    name="mass_1",
+    call="quantities::mass({df}, {output}, 0, {input})",
+    input=[q.ditaupair, q.Muon_mass],
+    output=[q.mass_1],
+    scopes=["mt"],
+)
+mass_2 = Producer(
+    name="mass_2",
+    call="quantities::mass({df}, {output}, 1, {input})",
+    input=[q.ditaupair, q.Tau_mass],
+    output=[q.mass_2],
+    scopes=["mt", "et", "tt"],
+)
+dxy_1 = Producer(
+    name="dxy_1",
+    call="quantities::dxy({df}, {output}, 0, {input})",
+    input=[q.ditaupair, q.Muon_dxy],
```

Comment:
I'd again suggest to hardcode such nanoAOD quantities in the call (unless certain difficulties arise, like a case we had with jets I think). Similar for most quantities below.

Reply:
yes, i will do this in a seperate PR for all existing producers at once